spec/unit/pops/parser/parser_rspec_helper.rb

```diff
@@ -4,6 +4,17 @@
 
 module ParserRspecHelper
   include FactoryRspecHelper
+
+  def with_app_management(flag)
+    # Horrible things have to be done to test this as the Lexer sets up the
+    # KEYWORD table as a frozen constant, and this list depends on the
+    # Puppet[:app_management] setting. These 'before all' and 'after all'
+    # clauses forces Ruby to reload the Lexer
+    Puppet[:app_management] = flag
+    Puppet::Pops::Parser.send(:remove_const, :Lexer2)
+    load 'puppet/pops/parser/lexer2.rb'
+  end
```

Comment:
Please add the following to ensure that the parser contained in the E4ParserAdapter is cleared if it exists.

```
Puppet::Parser::E4ParserAdapter.class_variable_set(:@@evaluating_parser, nil) if Puppet::Parser.const_defined?(:E4ParserAdapter)`
```
Also, please change the code in the _pops/parser/lexer2_spec.rb_ under "when app_management is (turned) on" so that it uses the `with_app_management(flag)` method.

This will take care of problems when executing tests in different order.